README.md
# revulyticsR
Connect to Your Revulytics Data with R!

## Purpose
revulyticsR facilitates making a connection to the Revulytics API and executing various queries. You can use it to get active users (daily, monthly, etc) or to query on various advanced events and get the results in tidy data frames.

## Installation
The development version can be installed from GitHub: `devtools::install_github("chrisumphlett/revulyticsR")`.

## Usage
A session must first be established before querying the API. This is done using your Revulytics username and password with `revultyics_auth()`.

The current version has several functions for making requests to the API.

* `get_active_users()`. For a given period of time (a day, week, or month) Revulytics' API summarizes and returns the number of active users. With this function you can return daily, weekly, or monthly active users for multiple product ids.
* `get_categories_and_events()`. For a list of product ids get all of the categories and events that have been defined (and identify it each is a basic or advanced). This can then be passed into subsequent queries to pull data on multiple events.
* `get_product_properties()` created. For a list of product ids get all of the product properties, both standard and custom. This can then be passed into `get_client_metadata()`.
* `get_client_metadata()` created. For a list of product ids get selected properties for each client, which is essentially metadata.  This works by pulling all of the clients that are installed within specified date range.

You will need your own credentials to use the package. A workflow could be:

```
  rev_user <- "my_username"
  rev_pwd <- "super_secret"
  product_ids_list <- c("123", "456", "789")
  start_date <- lubridate::floor_date(Sys.Date(), unit = "months") - months(6)
  end_date <- Sys.Date() - 1
  session_id <- revulytics_auth(rev_user, rev_pwd)
  monthly_active_users <- get_active_users(product_ids_list, "month", start_date, end_date, session_id, rev_user)
  category_event <- get_categories_and_events(product_ids_list, session_id, rev_user)
  product_properties <- get_product_properties(product_ids_list, session_id, rev_user)
  client_metadata <- get_client_metadata(product_ids_list, session_id, rev_user, product_properties, c("Property1", "Property2"), start_date, end_date)
```

More info on the API is available at https://docs.flexera.com/ui551/report/Content/helplibrary/APIRoot.htm#product-intro_184013195_1066562.

## Revulytics is now known as Flexera Usage Intelligence

While I was developing the package Revulytics was acquired by Flexera, but retained its name. When I was virtually done with the package Flexera rebranded it as Flexera Usage Intelligence. I expect that Revulytics is what it will still be commonly called for some time by its customers (as it is at my company).

NEWS.md
# revulyticsR 0.0.1

* `get_active_users()` created. For a given period of time (a day, week, or month) Revulytics' API summarizes and returns the number of active users. With this function you can return daily, weekly, or monthly active users for multiple product ids.
* `get_categories_and_events()` created. For a list of product ids get all of the categories and events that have been defined (and identify it each is a basic or advanced). This can then be passed into subsequent queries to pull data on multiple events.
* `get_product_properties()` created. For a list of product ids get all of the product properties, both standard and custom. This can then be passed into `get_client_metadata()`.
* `get_client_metadata()` created. For a list of product ids get selected properties for each client, which is essentially metadata.  This works by pulling all of the clients that are installed within specified date range.

# revulyticsR 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.
* Created initial package structure.
* Authentication/login method created.

R/revulytics_auth.R
#' Login and Obtain Revultyics API Session Id
#' 
#' A session must first be established before querying the API.
#' This is done using your Revulytics username and password.
#' 
#' It is not recommended that these values be stored directly 
#' in your code. There are various methods and packages 
#' available that are more secure; this package does not require
#' you to use any one in particular.
#' 
#' @param rev_username Revulytics username.
#' @param rev_password Revultyics password.
#' 
#' @import httr
#' @importFrom magrittr "%>%"
#' 
#' @return A list with details on connection to the Revulytics API.
#' 
#' @export
#' 
#' @examples
#' \dontrun{
#' rev_user <- "my_username"
#' rev_pwd <- "super_secret"
#' product_ids_list <- c("123", "456", "789")
#' start_date <- lubridate::floor_date(Sys.Date(), unit = "months") - months(6)
#' end_date <- Sys.Date() - 1
#' session_id <- revulytics_auth(rev_user, rev_pwd)
#' }

revulytics_auth <- function(rev_username, rev_password) {
  revulytics_login <- httr::POST("https://api.revulytics.com/auth/login",
                                 body = list(user = rev_username,
                                             password = rev_password,
                                             useCookies = FALSE),
                                 encode = "json")
  
  check_status(revulytics_login)
  
  rev_session_id <- httr::content(revulytics_login)$sessionId
  
  return(rev_session_id)
}



R/get_product_properties.R
#' Get All Properties for a List of Product Ids
#' 
#' Returns all of the unique properties (standard and custom)
#' for each product id by property category.
#' 
#' It is not recommended that your username be stored directly in your
#' code. There are various methods and packages available that are more 
#' secure; this package does not require you to use any one in particular.
#' 
#' @param rev_product_ids A vector of revulytics product id's for which
#' you want active user data.
#' @param rev_session_id Session ID established by the connection to
#' Revulytics API. This can be obtained with revulytics_auth().
#' @param rev_username Revulytics username.
#' 
#' @import dplyr
#' @importFrom magrittr "%>%"
#' @importFrom purrr "map_dfr"
#' @importFrom purrr "map_df"
#' @import httr
#' @import jsonlite
#' 
#' @return Data frame with properties and property attributes by
#' product id.
#' 
#' @export
#' 
#' @examples
#' \dontrun{
#' rev_user <- "my_username"
#' rev_pwd <- "super_secret"
#' product_ids_list <- c("123", "456", "789")
#' session_id <- revulytics_auth(rev_user, rev_pwd)  
#' product_properties <- get_product_properties(product_ids_list, session_id, rev_user)
#' }

get_product_properties <- function(rev_product_ids, rev_session_id, rev_username) {
  
  get_one_product_properties <- function(x){
    request_body <- list(
      user = rev_username,
      sessionId = rev_session_id,
      productId = x
    )
    
    body <- jsonlite::toJSON(request_body, auto_unbox = TRUE)
    request <- httr::POST("https://api.revulytics.com//meta/productProperties",
                           body = body,
                           encode = "json")
    check_status(request)
    
    request_content <- httr::content(request, "text", encoding = "ISO-8859-1")
    content_json <- jsonlite::fromJSON(request_content, flatten = TRUE)
    
    build_data_frame <- function(y){
      properties_category <- content_json$results$category[y]
      properties <- as.data.frame(content_json$results$properties[y]) %>%
        cbind(properties_category) %>%
        mutate(properties_category = as.character(.data$properties_category),
               revulytics_product_id = x,
               property_name = as.character(.data$name),
               property_friendly_name = as.character(.data$friendlyName),
               filter_type = as.character(.data$filterType),
               data_type = as.character(.data$dataType),
               supports_regex_f = as.character(if_else(.data$supportsRegex, 1, 0)),
               supports_meta_f = as.character(if_else(.data$supportsMeta, 1, 0)),
               supports_null_f = as.character(if_else(.data$supportsNull, 1, 0))) %>%
        select(.data$revulytics_product_id, .data$properties_category, .data$property_name, .data$property_friendly_name, .data$filter_type, 
               .data$data_type, .data$supports_regex_f, .data$supports_meta_f, .data$supports_null_f)

    }
    
    product_df <- map_df(1:length(content_json$results$category), build_data_frame)
    
  }
  
  all_products_df <- map_dfr(rev_product_ids, get_one_product_properties)
  return(all_products_df)
    
}
  
  




R/get_categories_and_events.R
#' Get All Categories and Events for a List of Product Ids
#' 
#' Returns all of the unique categories and events (basic and advanced)
#' for each product id.
#' 
#' It is not recommended that your username be stored directly in your
#' code. There are various methods and packages available that are more 
#' secure; this package does not require you to use any one in particular.
#' 
#' @param rev_product_ids A vector of revulytics product id's for which
#' you want active user data.
#' @param rev_session_id Session ID established by the connection to
#' Revulytics API. This can be obtained with revulytics_auth().
#' @param rev_username Revulytics username.
#' 
#' @import dplyr
#' @importFrom magrittr "%>%"
#' @importFrom purrr "map_dfr"
#' @import httr
#' @import jsonlite
#' 
#' @return Data frame with categories, events and event type by product id.
#' 
#' @export
#' 
#' @examples
#' \dontrun{
#' rev_user <- "my_username"
#' rev_pwd <- "super_secret"
#' product_ids_list <- c("123", "456", "789")
#' session_id <- revulytics_auth(rev_user, rev_pwd)  
#' category_event <- get_categories_and_events(product_ids_list, session_id, rev_user)
#' }

get_categories_and_events <- function(rev_product_ids, rev_session_id, rev_username) {

  get_by_product <- function(x) {
  
    request_body <- list(
      user = rev_username,
      sessionId = rev_session_id,
      productId = x,
      showEvents = array("all")
    )
    
    body <- jsonlite::toJSON(request_body, auto_unbox = TRUE)
    request <- httr::POST("https://api.revulytics.com/eventTracking/listEventNames",
                          body = body,
                          encode = "json")
    
    check_status(request)
    
    request_content <- httr::content(request, "text", encoding = "ISO-8859-1")
    content_json <- jsonlite::fromJSON(request_content, flatten = TRUE)
    
    parse_json_into_df <- function(x){
      category_name <- content_json$results$category[x]
      event_name <- as.data.frame(content_json$results$categoryEventNames[x]) %>%
        cbind(category_name) %>%
        mutate(category_name = as.character(.data$category_name))
    }
    
    category_event <- purrr::map_dfr(1:length(content_json$results$category), parse_json_into_df) %>%
      mutate(event_type = case_when(
                            .data$advanced ~ "ADVANCED",
                            .data$basic ~ "BASIC",
                            TRUE ~ "INACTIVE"),
             date_first_seen = as.Date(.data$dateFirstSeen),
             revulytics_product_id = x) %>%
      select(.data$revulytics_product_id, .data$category_name, .data$eventName, .data$event_type, .data$date_first_seen) %>%
      rename(event_name = .data$eventName)
    
  }
  
  category_event_by_prod <- purrr::map_dfr(rev_product_ids, get_by_product)
  return(category_event_by_prod)
}



R/get_active_users.R
#' Get Active Users by Product ID and Various Date Spans
#' 
#' For a given period of time (a day, week, or month) Revulytics' API
#' summarizes and returns the number of active users. With this function
#' you can return daily, weekly, or monthly active users for multiple 
#' product ids. 
#' 
#' You can specify a start and end date but Revulytics does not store
#' an indefinite period of historical data. In my experience this is 
#' three years but I do not know if this varies on a product or client
#' level.
#' 
#' It is not recommended that your username be stored directly in your
#' code. There are various methods and packages available that are more 
#' secure; this package does not require you to use any one in particular.
#' 
#' @param rev_product_ids A vector of revulytics product id's for which
#' you want active user data.
#' @param rev_date_type Level of aggregation, Revulytics will accept
#' "day", "week", or "month".
#' @param rev_start_date Date formatted YYYY-MM-DD. Revulytics may give
#' an error if you try to go back too far.
#' @param rev_end_date Date formatted YYYY-MM-DD.
#' @param rev_session_id Session ID established by the connection to
#' Revulytics API. This can be obtained with revulytics_auth().
#' @param rev_username Revulytics username.
#' 
#' @import dplyr
#' @importFrom magrittr "%>%"
#' @importFrom purrr "map_dfr"
#' @import httr
#' @import jsonlite
#' 
#' @return Data frame with active users for each product id and
#' unique date within the range
#' 
#' @export
#' 
#' @examples
#' \dontrun{
#' rev_user <- "my_username"
#' rev_pwd <- "super_secret"
#' product_ids_list <- c("123", "456", "789")
#' start_date <- lubridate::floor_date(Sys.Date(), unit = "months") - months(6)
#' end_date <- Sys.Date() - 1
#' session_id <- revulytics_auth(rev_user, rev_pwd)  
#' monthly_active_users <- get_active_users(product_ids_list,
#' "month",
#' start_date,
#' end_date,
#' session_id,
#' rev_user)
#' }


get_active_users <- function(rev_product_ids, rev_date_type, rev_start_date, rev_end_date, rev_session_id, rev_username) {
  
  . <- NA # prevent variable binding note for the dot in the get_by_product function
  
  get_by_product <- function(x, rev_date_type) {
    request_body <- list(
      user = rev_username,
      sessionId = rev_session_id,
      productId = x,
      startDate = rev_start_date,
      stopDate = rev_end_date,
      dateSplit = rev_date_type,
      clientStatus = array("active")
    )
    
    request <- httr::POST("https://api.revulytics.com/reporting/generic/dateRange?responseFormat=raw",
                     body = request_body,
                     encode = "json")
    
    check_status(request)
    
    request_content <- httr::content(request, "text", encoding = "ISO-8859-1")
    content_json <- jsonlite::fromJSON(request_content, flatten = TRUE)
    
    iteration_df <- as.data.frame(unlist(content_json$results)) %>% cbind(rownames(.)) %>%
      dplyr::rename(active_user_date = 2, active_users = 1) %>%
      dplyr::mutate(active_user_date = as.Date(substr(.data$active_user_date, 1, 10)),
             revulytics_product_id = x)
    rownames(iteration_df) <- NULL
    return(iteration_df)
  }
  
  df <- purrr::map_dfr(rev_product_ids, get_by_product, rev_date_type)
  
  return(df)
}

